### PR TITLE
WIP: add installation for Intel oneAPI icc/icpc/ifort

### PR DESCRIPTION
**Pull request by rscohn2**
PR is for illustration, not intended for merging.

This will install icc, icpc, and ifort from intel oneapi beta 8. 

You have to source setvars.sh in the install root to set up the path to use the compilers.

I tested the command line in isolation, but not as part of the script.

### CODE REVIEW

**Comment by partouf** [comment]
Hi, sorry I haven't had the chance to look at this. But thank you for doing this. I will look at this somewhere next week if no one else does first.

**Comment by rscohn2** [comment]
No problem. Let me know if there is more I can do to help.

This will install the oneAPI 2021.1-beta08 icc and fortran compiler. oneAPI 2021.1-beta08 is very close to parallel studio 19.1. I gave you a oneapi install instead of parallel studio because it is simpler from the command line and the future direction.

If you are getting requests for the parallel studio compiler, then I could give you something similar if you can host your license file on s3.

**Comment by mattgodbolt** [comment]
Thanks @rscohn2 ! The intel installation is always fraught with trickiness! It has definitely caused us trouble before: some intel engineers have sent me similar things before to get installs working...

File: update_compilers/install_nonfree_compilers.sh [hunk 1]
```diff
@@ -36,6 +36,20 @@ for version in 2016.3.210 2018.0.033; do
...
+    local HPCKIT_PATH=${TMP_DIR}/hpckit.sh
+    local INSTALL_DIR=$(pwd)/$2
+    mkdir -p ${TMP_DIR}
+    fetch ${HPCKIT_URL} > $HPCKIT_PATH
+    bash $HPCKIT_PATH -s -a -s --action install --eula accept --components intel.oneapi.lin.cpp-compiler:intel.oneapi.lin.ifort-compiler --install-dir ${DIR}
+    # to use the compiler, you must 'source ${DIR}/setvars.sh'
+    rm -rf ${TMP_DIR}
+}
+
+install_oneapi https://registrationcenter-download.intel.com/akdlm/irc_nas/16856/l_HPCKit_b_2021.1.8.1883_offline.sh 2021-beta8
```
Comment (mattgodbolt) [inline update_compilers/install_nonfree_compilers.sh:51]: This is great: how likely is the download to stick around? My experience is that "normally" versioned things like "gcc-9.2.0" the tarballs tend to stay around forever, but relatively unusual seemingly random things like this tend to vanish from download servers randomly.

If the latter is true we may wish to do what we "normally" do; which is to install, then tarball up and save the resulting installation image. This is how the existing intel compilers are installed.

Basically this comes down to a kind of "backup", can we recover if our NFS drive gets zapped my a lightning bolt. Or idiocy on my part :)  Perhaps I'm just paranoid!

Comment (rscohn2) [reply update_compilers/install_nonfree_compilers.sh:51]: The URLs do not change. The 5 digit number is the unique identifier for the release in the content management system.

It looks like there is some manual effort for you to post a new compiler. I was hoping we could get to a point where we could just submit PR's when new compilers are available.

Comment (partouf) [reply update_compilers/install_nonfree_compilers.sh:51]: I'd say let's use this new way (fitted into the installer system and not this script) and blame Intel when things go missing

Comment (mattgodbolt) [reply update_compilers/install_nonfree_compilers.sh:51]: Thanks both: agreed! I think this is a big improvement!

I'm in the process of changing how we install things: I can ensure it works for this too.

Comment (mattgodbolt) [reply update_compilers/install_nonfree_compilers.sh:51]: Re: `# to use the compiler, you must 'source ${DIR}/setvars.sh'` -- we'll have to find a way around this as we don't run a regular shell to do compilations.

Comment (rscohn2) [reply update_compilers/install_nonfree_compilers.sh:51]: This works for me:

LD_LIBRARY_PATH=/opt/intel/oneapi/compiler/latest/linux/compiler/lib/intel64_lin/:$LD_LIBRARY_PATH /opt/intel/oneapi/compiler/latest/linux/bin/intel64/icpc  ex.cpp

**Comment by partouf** [comment]
While downloading and trying to install this, I'm reading this on the website https://dynamicinstaller.intel.com/oneapi/toolkits/hpc-kit/linux/
`I understand that the full functionality of Intel® oneAPI HPC Toolkit requires the Intel® oneAPI Base Toolkit to be previously installed. By proceeding with the Intel® oneAPI HPC Toolkit without a previously installed Intel® oneAPI Base Toolkit means that some functionality will not be present. Check the box to enable Intel® oneAPI HPC Toolkit download.`

I don't know whether or not we want/need the Base toolkit thing, not sure what it means.

**Comment by partouf** [comment]
These might be what's in the Base https://software.intel.com/content/www/us/en/develop/articles/intel-oneapi-toolkit-release-notes.html

We use TBB, but we build it from the github repo, so I think we're good.

**Comment by partouf** [comment]
Now considering the sluglike speed of the download, maybe storing it somewhere else is preferable.

**Comment by rscohn2** [comment]
We recently setup 6 different CI systems for oneapi: [https://github.com/oneapi-src/oneapi-ci](https://github.com/oneapi-src/oneapi-ci). You can see it run with github actions here: https://github.com/oneapi-src/oneapi-ci/runs/1128321693?check_suite_focus=true

Linux installs icc, ifort, & dpcpp with apt (2-3 minutes). Windows uses our binary installer (< 3 minutes).

APT allows you to install multiple versions of the compiler under a common root: /opt/intel/oneapi. If you are OK with not being able to change the root, then apt might be better. If you want to control the install root, then you need the binary installer.

The binary installer has an online mode with a small initial download, and then downloads the components it needs. I gave you the offline mode, where you gave a very large initial download, and it will not access the network during install.

I gave you the binary installer with offline because it is the most flexible. If you would rather use apt or binary installer in online mode, I can give you directions. It works almost the same as offline, but will be faster if you are only installing the compiler.

**Comment by partouf** [comment]
> APT allows you to install multiple versions of the compiler under a common root: /opt/intel/oneapi. If you are OK with not being able to change the root, then apt might be better. If you want to control the install root, then you need the binary installer.

We install the compilers on an EFS drive so all our scaling instances can access the compilers, so not physically on the actual system that it will use the installed compilers.

I managed to install and able to compile, build and execute with the version linked here.
Using this installer https://github.com/compiler-explorer/infra/compare/intelcompilerthings#diff-d6522969e260fca85915c55f6adc4c3fR349

But when I tried installing the one I could download today (2021.1.9.2205), the installer returns an errorcode, but the logging doesn't seem to indicate any kind of error.

```
2020-09-17 14:49:41,609 lib.installation INFO     compilers/c++/icc 2021.1.9.2205: /home/partouf/install_intel.sh -> install.sh
2020-09-17 14:49:41,610 lib.installation INFO     Staging with bash -c rm -Rf ~/.intel
rm -Rf ~/intel
bash install.sh -s -a -s --action install --eula accept --components intel.oneapi.lin.cpp-compiler:intel.oneapi.lin.ifort-compiler --install-dir /opt/compiler-explorer/staging/intel-2021.1.9.2205
rm -Rf ~/.intel
rm -Rf ~/intel

Checking system requirements...
Done.
Wait while preparing installer...
Done.
Launching installer...
Failed to select components to install
Installer completed with code 3
Log file: /tmp/partouf/intel_oneapi_installer/2020.09.17_14.50.09.083/installer.install.intel.oneapi.lin.hpckit.product,v=2021.1-2205.beta09.2020.09.17_14.50.11.414.1125680332.log
```

Contents of logfile:
```
09/17/2020 14:50:12:025 : 007520 : MESSAGE : /home/partouf/intel/oneapi/installer/installer is run with arguments: -s --action install --eula accept --components intel.oneapi.lin.cpp-compiler:intel.oneapi.lin.ifort-compiler --install-dir /opt/compiler-explorer/staging/intel-2021.1.9.2205 --log-dir /tmp/partouf/intel_oneapi_installer/2020.09.17_14.50.09.083 --product-id intel.oneapi.lin.hpckit.product --product-ver 2021.1-2205.beta09 --package-path /tmp/tmp.988xGj7akU/l_HPCKit_b_2021.1.9.2205_offline/packages
09/17/2020 14:50:12:025 : 007520 : MESSAGE : Installer is started in silent mode
09/17/2020 14:50:12:025 : 007520 : MESSAGE : Installer executable location is /home/partouf/intel/oneapi/installer/installer, version is 4.0.2-162.beta
09/17/2020 14:50:12:025 : 007520 : MESSAGE : Package manager executable location is /home/partouf/intel/packagemanager/1.0/packagemanager, version is 4.0.0-104.beta
09/17/2020 14:50:12:025 : 007520 : MESSAGE : Product is set to intel.oneapi.lin.hpckit.product,v=2021.1-2205.beta09. Mode is set to install.
09/17/2020 14:50:12:114 : 007523 : MESSAGE : PreRequisite for other install successfully checked
09/17/2020 14:50:12:121 : 007523 : MESSAGE : PreRequisite for other Intel install successfully checked
09/17/2020 14:50:12:125 : 007523 : MESSAGE : All checks for preRequisite for supported operation system are passed
```

**Comment by partouf** [comment]
The latest installer doesn't seem to contain the C++ compiler? Or is it combined into intel.oneapi.lin.dpcpp-cpp-compiler-pro?

```
partouf@ceub20:/opt/compiler-explorer/infra$ bash /opt/compiler-explorer/staging/install.sh -a --list-components
Extract l_HPCKit_b_2021.1.9.2205_offline to /tmp/tmp.biVVV3lS8J/l_HPCKit_b_2021.1.9.2205_offline...
[########################################################################################################################################################################################################################]
Extract l_HPCKit_b_2021.1.9.2205_offline completed!
Checking system requirements...
Done.
Wait while preparing installer...
Done.
Launching installer...
ID                                      Version            Language Name                                 
=========================================================================================================
intel.oneapi.lin.dpcpp-cpp-compiler-pro 2021.1-2198.beta09          Intel® oneAPI DPC++/C++ Compiler Pro 
intel.oneapi.lin.ifort-compiler         2021.1-2198.beta09          Intel® Fortran Compiler Classic      
intel.oneapi.lin.mpi.devel              2021.1-2099.beta09          Intel® MPI Library                   
intel.oneapi.lin.inspector              2021.1-319.beta09           Intel® Inspector                     
intel.oneapi.lin.clck                   2021.1-760.beta09           Intel® Cluster Checker               
intel.oneapi.lin.itac                   2021.1-376.beta09           Intel® Trace Analyzer and Collector  
Remove extracted files: [/tmp/tmp.biVVV3lS8J/l_HPCKit_b_2021.1.9.2205_offline] ...

```

**Comment by partouf** [comment]
Ok, so it's now in the dpcpp-cpp-compiler-pro component it seems. It works if I install that one instead.

**Comment by rscohn2** [comment]
For a new release, will updating the yaml be everything that is needed to make a new compiler available to users?

**Comment by partouf** [comment]
> For a new release, will updating the yaml be everything that is needed to make a new compiler available to users?

Pretty much, the compilers have to be made available to CE itself as well https://github.com/compiler-explorer/compiler-explorer/pull/2208/files

And then we run the installer manually on our administrative AWS instance, or we wait a day for it to automatically run. And then we push the changes live.

**Comment by partouf** [comment]
C++ Libraries have built for the 2 new compilers without problems as far as I can see. So they are ready to go live.

Note btw that I didn't add ispc via this route, because we already build the trunk nightly from the github repo. I think it's just missing the 1.14.1 tagged version at the moment. Might add that later.

**Comment by rscohn2** [comment]
We don't distribute ispc through these installers, building from the repo is the right thing to do. Or maybe I am misunderstanding your reference to ispc?

There is a desire to add dpcpp, but someone else was looking into it. Extracting the generated code is different because it produces fat binaries that target GPU and x86.

**Comment by partouf** [comment]
> We don't distribute ispc through these installers, building from the repo is the right thing to do. Or maybe I am misunderstanding your reference to ispc?
> 
> There is a desire to add dpcpp, but someone else was looking into it. Extracting the generated code is different because it produces fat binaries that target GPU and x86.

Ah right, I got confused because you initially mentioned icpc - I guess I don't understand what that is then. I confused it with ispc.

**Comment by rscohn2** [comment]
icc is comparable to gcc, and icpc is the g++ equivalent.

**Comment by partouf** [comment]
These changes and the compilers are now live on the website
